clang/lib/Format/TokenAnnotator.cpp

```diff
@@ -4591,7 +4589,9 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
     if (!BeforeLeft)
       return false;
     if (BeforeLeft->is(tok::coloncolon)) {
-      return Left.is(tok::star) &&
+      const auto *Prev = BeforeLeft->Previous;
+      return Left.is(tok::star) && Prev &&
+             !Prev->endsSequence(tok::identifier, TT_FunctionTypeLParen) &&
```

Comment:
i think we can also have cases like `void (A::B::*func)();`

Reply:
I had thought about adding a loop to handle the nested-name-specifier but decided to wait until we ran into code like the example you gave above in practice. Let me know if you strongly feel that we should handle it now.

Reply:
i think it's actually common enough to handle in this patch, https://github.com/search?q=lang%3Acpp+%2F%3A%3A%5Cw%2B%3A%3A%5C*%5C%29%2F&type=code.

moreover we also have cases like `(std::remove_pointer<O>::type::*)`. so it might actually be easier to match by going forward (which should also ensure we do less operations)? e.g. something like:
```
// Don't add space around function-type pointers.
// - void (*foo)(int);
// - void (::B::*foo)(int);
// - void (A<T>::B::*)(int);
auto OpeningParen = Right.MatchingParen;
if (!OpeningParen && Right.Next)
  OpeningParen = Right.Next->MatchingParen;
if (OpeningParen && OpeningParen->is(TT_FunctionTypeLParen))
  return false;

if (BeforeLeft->is(tok::coloncolon))
  return true;
```